test/tcp_echo_client/tcp_echo_client.cpp

```diff
@@ -20,17 +20,201 @@
 #include "sal-stack-lwip/lwipv4_init.h"
 #include "EthernetInterface.h"
 #include "mbed-drivers/test_env.h"
+#include "sal/socket_api.h"
 
-void app_start(int argc, char *argv[])
+#ifndef TCP_EC_SOCKET_TEST_TIMEOUT
+#define TCP_EC_SOCKET_TEST_TIMEOUT 1.0f
+#endif
+
+
+static struct socket *tcp_ec_client_socket_g;
+static volatile bool tcp_ec_client_event_done_g;
+static volatile bool tcp_ec_client_rx_done_g;
+static volatile bool tcp_ec_client_tx_done_g;
+static volatile struct socket_tx_info tcp_ec_client_tx_info_g;
+static volatile struct socket_event tcp_ec_client_event_g;
+static volatile int tcp_ec_closed;
+static volatile int tcp_ec_connected;
+
+volatile int tcp_ec_timeout_g;
+static void onTimeout() {
+    tcp_ec_timeout_g = 1;
+}
+
+
+/*****************************************************************************
+ * FUNCTION: tcp_ec_sock_addr_port_dump
+ *  pretty print a socket address and port with description
+ * ARGUMENTS:
+ *  description     client string to provide contextural information for
+ *                  the ipaddr:port tuple
+ *  addr            ip address data structure
+ *  port            port number
+ * RETURN:
+ *  None
+ *****************************************************************************/
+static inline void tcp_ec_sock_addr_port_dump(const char* description, struct socket_addr* saddr, uint16_t port)
+{
+    if(saddr)
+    {
+        TEST_PRINT("%s:%s:%d\r\n", description, inet_ntoa(*saddr), port);
+    }
+}
+
+/*****************************************************************************
+ * FUNCTION: tcp_ec_client_cb
+ * callback function for handling tx/rx/etc event indications.
+ * script
+ * ARGUMENTS:
+ *  None
+ * RETURN:
+ *  None
+ *****************************************************************************/
+static void tcp_ec_client_cb()
+{
+    struct socket_event *e = tcp_ec_client_socket_g->event;
+    event_flag_t event = e->event;
+    switch (event) {
+        case SOCKET_EVENT_RX_DONE:
+            tcp_ec_client_rx_done_g = true;
+            break;
+        case SOCKET_EVENT_TX_DONE:
+            tcp_ec_client_tx_done_g = true;
+            tcp_ec_client_tx_info_g.sentbytes = e->i.t.sentbytes;
+            break;
+        case SOCKET_EVENT_DISCONNECT:
+            tcp_ec_closed = 1;
+            break;
+        case SOCKET_EVENT_CONNECT:
+            tcp_ec_connected = 1;
+            break;
+        default:
+            memcpy((void *) &tcp_ec_client_event_g, (const void*)e, sizeof(e));
```

Comment:
If the test inspects the received event and processes it correctly later, that's fine. I just wanted to know whether it would make more sense to trap a failure here, in IRQ context, or whether it should be at a higher level. If the higher level code is in place, that's great!